pkg/controllers/provisioning/provisioner.go

```diff
@@ -245,46 +251,68 @@ func (p *Provisioner) launch(ctx context.Context, node *scheduling.Node) error {
 		}
 	}
 	logging.FromContext(ctx).Infof("Created %s", node)
-	if err := p.bind(ctx, k8sNode, node.Pods); err != nil {
-		return fmt.Errorf("binding pods, %w", err)
+	for _, pod := range node.Pods {
+		p.recorder.PodShouldSchedule(pod, k8sNode)
 	}
 	return nil
 }
 
-func (p *Provisioner) bind(ctx context.Context, node *v1.Node, pods []*v1.Pod) (err error) {
-	defer metrics.Measure(bindTimeHistogram.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
-
-	nodeTaints := v1alpha5.Taints(node.Spec.Taints)
-
-	notReadyTolerations := []v1.Toleration{
-		{
-			Key:      v1alpha5.NotReadyTaintKey,
-			Operator: v1.TolerationOpEqual,
-			Effect:   v1.TaintEffectNoSchedule,
-		}, {
-			Key:      v1.TaintNodeNotReady,
-			Operator: v1.TolerationOpEqual,
-			Effect:   v1.TaintEffectNoSchedule,
-		}}
-
-	workqueue.ParallelizeUntil(ctx, len(pods), len(pods), func(i int) {
-		pod := pods[i]
-		// Don't bind pods that would immediately get evicted.  We tolerate the two standard taints that are applied for
```

Comment:
I've removed binding entirely here, but there are other options that are probably worth discussing.

Reply:
Did you have other options in mind?  Would knowing what node we think the pod will schedule to help in anyway?

Reply:
Other options:
1) Stop pre-binding only for nodes with extended resources (should fix both GPU and SGPP issues)
2) We could bind pods as a last step before removing the `karpenter.sh/not-ready` taint.
3) We could 'nominate a node' for pods as a last step before removing the `karpenter.sh/not-ready` taint.

Reply:
> 1. Stop pre-binding only for nodes with extended resources (should fix both GPU and SGPP issues)

only pre-binding for extended resources would be surprising I would think. 

> 2. We could bind pods as a last step before removing the karpenter.sh/not-ready taint.

This would give us more guarantees if there were a lot of pods coming in that had various resource requests that would throw off proper bin-packing. This would be less of a change too. 

> 3. We could 'nominate a node' for pods as a last step before removing the karpenter.sh/not-ready taint.

Would this really give us anything? Wouldn't opt 2 be the same thing but a stronger form.



It seems to me that option 2 is the most sensible at this point. We fix the issues with extended-resources, Karpenter behaves in roughly the same way of ultimately scheduling the pod reducing the chance of fragmentation from an influx of pods that karpenter is unaware of,  and works the same for all pods. We could always completely remove binding later if we found a "late pre-binding" not needed.

Reply:
> > 2. We could bind pods as a last step before removing the karpenter.sh/not-ready taint.
> 
> This would give us more guarantees if there were a lot of pods coming in that had various resource requests that would throw off proper bin-packing. This would be less of a change too.
> 
> > 3. We could 'nominate a node' for pods as a last step before removing the karpenter.sh/not-ready taint.
> 
> Would this really give us anything? Wouldn't opt 2 be the same thing but a stronger form.
> 
> It seems to me that option 2 is the most sensible at this point. We fix the issues with extended-resources, Karpenter behaves in roughly the same way of ultimately scheduling the pod reducing the chance of fragmentation from an influx of pods that karpenter is unaware of, and works the same for all pods. We could always completely remove binding later if we found a "late pre-binding" not needed.

Yeah, 2 and 3 were similar.  After further thought, they may both be broken concepts though.  Device plugin daemonsets which create the extended resources may or may not (e.g. nvidia won't) tolerate our taint.  If they don't, then they won't bind and won't add the extended resources.  

Maybe by the time the node is ready, Kubelet won't zero out the resources so we can optimistically bind the pod and then remove our taint which would allow the DS to schedule.  However, device plugins potentially could zero out resources then which would cause pods to get evicted.